src/java_bytecode/java_bytecode_convert_method.cpp

```diff
@@ -556,12 +556,211 @@ codet java_bytecode_convert_methodt::get_array_bounds_check(
   bounds_checks.operands().back().add_source_location()
     .set_property_class("array-index-out-of-bounds-high");
 
-  // TODO make this throw ArrayIndexOutOfBoundsException instead of asserting.
   return bounds_checks;
 }
 
 /*******************************************************************\
 
+Function: throw_array_access_exception
+
+  Inputs: 
+
+ Outputs: 
+
+ Purpose: Checks whether the array access array_struct[idx] 
+          is out-of-bounds, and throws ArrayIndexOutofBoundsException 
+          if necessary
+
+\*******************************************************************/
+
+codet java_bytecode_convert_methodt::throw_array_access_exception(
+  const exprt &array_struct,
+  const exprt &idx,
+  const source_locationt &original_loc)
+{
+  symbolt exc_symbol;
+  if(!symbol_table.has_symbol("ArrayIndexOutOfBoundsException"))
+  {
+    exc_symbol.is_static_lifetime=true;
+    exc_symbol.base_name="ArrayIndexOutOfBoundsException";
+    exc_symbol.name="ArrayIndexOutOfBoundsException";
+    exc_symbol.mode=ID_java;
+    exc_symbol.type=typet(ID_pointer, empty_typet());
+    symbol_table.add(exc_symbol);
+  }
+  else
+    exc_symbol=symbol_table.lookup("ArrayIndexOutOfBoundsException");
+
+  const symbol_exprt &exc=exc_symbol.symbol_expr();
+  side_effect_expr_throwt throw_expr;
+  throw_expr.add_source_location()=original_loc;
+  throw_expr.copy_to_operands(exc);
+
+  const constant_exprt &zero=from_integer(0, java_int_type());
+  const binary_relation_exprt lt_zero(idx, ID_lt, zero);
+  const member_exprt length_field(array_struct, "length", java_int_type());
+  const binary_relation_exprt ge_length(idx, ID_ge, length_field);
+  const and_exprt and_expr(lt_zero, ge_length);
```

Comment:
Should be `or_exprt` I think? I believe this is testing for `idx < 0 && idx >= arr.length`